src/MonoTorrent/MonoTorrent.Client/Managers/TrackerManager.cs

```diff
@@ -108,80 +108,114 @@ internal TrackerManager (ITrackerRequestFactory requestFactory, IEnumerable<RawT
-                try {
-                    // If we have not announced to this Tracker tier yet then we should replace the ClientEvent.
-                    // But if we end up announcing to a different Tracker tier we may want to send the
-                    // original/unmodified args.
-                    AnnounceParameters actualArgs = p;
-                    if (!trackerTier.SentStartedEvent)
-                        actualArgs = actualArgs.WithClientEvent (TorrentEvent.Started);
-
-                    List<Peer> peers = await tracker.AnnounceAsync (actualArgs);
-                    LastAnnounceSucceeded = true;
-                    AnnounceComplete?.InvokeAsync (this, new AnnounceResponseEventArgs (tracker, true, peers.AsReadOnly ()));
+        async Task AnnounceToTier (TorrentEvent clientEvent, TrackerTier tier)
+        {
+            for (int i = 0; i < tier.Trackers.Count; i++) {
+                int trackerIndex = (i + tier.ActiveTrackerIndex) % tier.Trackers.Count;
+                var tracker = tier.Trackers[trackerIndex];
+
+                var result = await Announce (clientEvent, tracker);
+                tier.LastAnnounce = ValueStopwatch.StartNew ();
+                if (result) {
+                    tier.ActiveTrackerIndex = trackerIndex;
+                    tier.LastAnnounceSucceeded = true;
                     return;
-                } catch {
                 }
             }
 
-            LastAnnounceSucceeded = false;
-            AnnounceComplete?.InvokeAsync (this, new AnnounceResponseEventArgs (null, false));
+            // All trackers failed to respond.
+            tier.LastAnnounceSucceeded = false;
+        }
+
+        async Task<bool> Announce (TorrentEvent clientEvent, ITracker tracker)
+        {
+            var trackerTier = Tiers.First (t => t.Trackers.Contains (tracker));
+            try {
+                // If we have not announced to this Tracker tier yet then we should replace the ClientEvent.
+                // But if we end up announcing to a different Tracker tier we may want to send the
+                // original/unmodified args.
+                AnnounceParameters actualArgs = RequestFactory.CreateAnnounce (clientEvent);
+                if (!trackerTier.SentStartedEvent)
+                    actualArgs = actualArgs.WithClientEvent (TorrentEvent.Started);
+
+                List<Peer> peers = await tracker.AnnounceAsync (actualArgs);
+                trackerTier.LastAnnounceSucceeded = true;
+
+                trackerTier.ActiveTrackerIndex = trackerTier.Trackers.IndexOf (tracker);
+                trackerTier.SentStartedEvent |= actualArgs.ClientEvent == TorrentEvent.Started;
+                trackerTier.LastAnnounce = ValueStopwatch.StartNew ();
+                AnnounceComplete?.InvokeAsync (this, new AnnounceResponseEventArgs (tracker, true, peers.AsReadOnly ()));
+                return true;
+            } catch {
```

Comment:
May I make a suggestion. Currently you have following TrackerStates:
    public enum TrackerState
    {
        Ok,
        Offline,
        InvalidResponse
    }

Could you also add Unknown and connecting Connecting.
Looks like here is the right place to set tracker state now.
All trackers should start with Unknown state, once you start connecting, you change it to Connencting.
If connection succeed, you change it to Ok, if fails - to Offline or InvalidResponse.

Reply:
Yup, makes sense. That should be easy to add!

Reply:
Tracking this here: https://github.com/alanmcgovern/monotorrent/issues/255